src/DSi.cpp

```diff
@@ -583,17 +622,29 @@ void MapNWRAM_A(u32 num, u8 val)
     MBK[0][mbkn] |= (val << mbks);
     MBK[1][mbkn] = MBK[0][mbkn];
 
-    u8* ptr = &NWRAM_A[num << 16];
-
-    if (oldval & 0x80)
+    // When we only update the mapping on the written MBK, we will
+    // have priority of the last witten MBK over the others
+    // However the hardware has a fixed order. Therefor 
+    // we need to iterate through them all in a fixed order and update
+    // the mapping, so the result is independend on the MBK write order
+    for (u8 part = 0; part < 4; part++)
     {
-        if (NWRAMMap_A[oldval & 0x01][(oldval >> 2) & 0x3] == ptr)
-            NWRAMMap_A[oldval & 0x01][(oldval >> 2) & 0x3] = NULL;
+        NWRAMMap_A[0][part] = NULL;
+        NWRAMMap_A[1][part] = NULL;
     }
-
-    if (val & 0x80)
+    // can only map if NWRAM is available
+    if (!NWRAM_A)
+        NWRAM_A = new u8[NWRAMSize];
```

Comment:
NWRAM_A should always be allocated here. Besides you can't just allocate heap memory here at all because when fast memory is enabled special pointers related to it are used for this.

Reply:
You are correct. I had issues with NWRAM_A/B/C beein empty in the process, but that later in debugging resolved as an out of bounds write to the MBK array.